www/class/config-generate/service.class.php

```diff
@@ -94,193 +103,377 @@ private function getServiceCache()
+            : null;
+        $serviceIdTopLevel = $serviceId;
+        $computedCache = array();
+        if (!is_null($serviceIdTopLevel) && !isset($servicesTpl[$serviceIdTopLevel][$attribute . '_computed_cache'])) {
+            $loop = array();
+            while (!is_null($serviceId)) {
+                if (isset($loop[$serviceId])) {
+                    break;
+                }
+                $loop[$serviceId] = 1;
+
+                if (!is_null($servicesTpl[$serviceId]['notifications_enabled'])
+                    && (int)$servicesTpl[$serviceId]['notifications_enabled'] === 0) {
+                    break;
+                }
+
+                if (count($servicesTpl[$serviceId][$attribute . '_cache']) > 0) {
+                    $computedCache = array_merge($computedCache, $servicesTpl[$serviceId][$attribute . '_cache']);
+                    if (is_null($servicesTpl[$serviceId][$attributeAdditive])
+                        || $servicesTpl[$serviceId][$attributeAdditive] != 1) {
+                        break;
+                    }
+                }
+                $serviceId = isset($servicesTpl[$serviceId]['service_template_model_stm_id'])
+                    ? $servicesTpl[$serviceId]['service_template_model_stm_id']
+                    : null;
+            }
+            $servicesTpl[$serviceIdTopLevel][$attribute . '_computed_cache'] = array_unique($computedCache);
         }
 
-        if ((isset($this->service_cache[$service_id]['contacts']) &&
-                $this->service_cache[$service_id]['contacts'] != '') ||
-            (isset($this->service_cache[$service_id]['contact_groups']) &&
-                $this->service_cache[$service_id]['contact_groups'] != '')
-        ) {
-            return 1;
+        if (!is_null($serviceIdTopLevel)) {
+            $results = array_unique(
+                array_merge($results, $servicesTpl[$serviceIdTopLevel][$attribute . '_computed_cache']),
+                SORT_NUMERIC
+            );
         }
+        return $results;
+    }
 
-        return 0;
+    /**
+     * @param array $service
+     * @param array|null $cg
+     */
+    private function setContactGroups(array &$service, ?array $cg) : void
+    {
+        $cgInstance = Contactgroup::getInstance($this->dependencyInjector);
+        $cgResult = '';
+        $cgResultAppend = '';
+        foreach ($cg as $cgId) {
+            $tmp = $cgInstance->generateFromCgId($cgId);
+            if (!is_null($tmp)) {
+                $cgResult .= $cgResultAppend . $tmp;
+                $cgResultAppend = ',';
+            }
+        }
+        if ($cgResult != '') {
+            $service['contact_groups'] = $cgResult;
+        }
     }
 
-    private function getContactsFromHost($host_id, $service_id, $sOnlyContactHost)
+    /**
+     * @param array $service
+     * @param array|null $contacts
+     */
+    private function setContacts(array &$service, ?array $contacts): void
```

Comment:
This array is used in a foreach loop so it cannot be null.
```suggestion
    private function setContacts(array &$service, array $contacts = []): void
```